test/lib/services/html_converter_test.rb

```diff
@@ -131,13 +131,100 @@ def test_transform_html_on_default_lang_with_query_pattern_and_supported_lang
       assert_equal(expected_html, translated_html)
     end
 
+    def test_transform_html_on_target_lang_translate_canonical_tag
+      dom = get_dom('<html><head><link rel="canonical" href="http://my-site.com/" /></head><body></body></html>')
+      settings = {
+        'default_lang' => 'en',
+        'supported_langs' => %w[en ja vi],
+        'url_pattern' => 'path',
+        'translate_canonical_tag' => true
+      }
+      store, headers = store_headers_factory(settings, 'http://my-site.com/vi/')
+      url_lang_switcher = Wovnrb::UrlLanguageSwitcher.new(store)
+      converter = HtmlConverter.new(dom, store, headers, url_lang_switcher)
+      translated_html = converter.build
+
+      expected_html = "<html lang=\"en\"><head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\"><script src=\"https://j.wovn.io/1\" async=\"true\" data-wovnio=\"key=123456&amp;backend=true&amp;currentLang=vi&amp;defaultLang=en&amp;urlPattern=path&amp;langCodeAliases={}&amp;langParamName=wovn&amp;version=WOVN.rb_#{VERSION}\" data-wovnio-type=\"fallback_snippet\"></script><link rel=\"canonical\" href=\"http://my-site.com/vi/\"><link rel=\"alternate\" hreflang=\"en\" href=\"http://my-site.com/\"><link rel=\"alternate\" hreflang=\"ja\" href=\"http://my-site.com/ja/\"><link rel=\"alternate\" hreflang=\"vi\" href=\"http://my-site.com/vi/\"></head><body></body></html>"
+      assert_equal(expected_html, translated_html)
+    end
+
+    def test_transform_html_on_default_lang_do_not_change_canonical_tag_path_pattern
+      dom = get_dom('<html><head><link rel="canonical" href="http://my-site.com/" /></head><body></body></html>')
+      settings = {
+        'default_lang' => 'en',
+        'supported_langs' => %w[en ja vi],
+        'url_pattern' => 'path',
+        'translate_canonical_tag' => true
+      }
+      store, headers = store_headers_factory(settings, 'http://my-site.com/')
+      url_lang_switcher = Wovnrb::UrlLanguageSwitcher.new(store)
+      converter = HtmlConverter.new(dom, store, headers, url_lang_switcher)
+      translated_html = converter.build
+
+      expected_html = "<html lang=\"en\"><head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\"><script src=\"https://j.wovn.io/1\" async=\"true\" data-wovnio=\"key=123456&amp;backend=true&amp;currentLang=en&amp;defaultLang=en&amp;urlPattern=path&amp;langCodeAliases={}&amp;langParamName=wovn&amp;version=WOVN.rb_#{VERSION}\" data-wovnio-type=\"fallback_snippet\"></script><link rel=\"canonical\" href=\"http://my-site.com/\"><link rel=\"alternate\" hreflang=\"en\" href=\"http://my-site.com/\"><link rel=\"alternate\" hreflang=\"ja\" href=\"http://my-site.com/ja/\"><link rel=\"alternate\" hreflang=\"vi\" href=\"http://my-site.com/vi/\"></head><body></body></html>"
+      assert_equal(expected_html, translated_html)
+    end
+
+    def test_transform_html_on_default_lang_do_not_change_canonical_tag_query_pattern
```

Comment:
i don't think the pattern is particularly relevant to the canonical tag feature

Reply:
The different patterns are included because canonical tag is using the new ported `UrlLanguageSwitcher`. Since it's now only used by canonical tag feature I feel it's ok to test it via canonical tag. If we decide to make use it for all URL swapping needs then we can have separate tests for them for sure.

Reply:
you may as well change the other usages to use this, no?